### PR TITLE
GH-100892: Fix race in clearing `threading.local`

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-100892

### CODE REVIEW

**Comment by kumaraditya303** [comment]
This race exists in security only branches too, @ambv  how about backporting this to those?

File: Modules/_threadmodule.c [hunk 1]
```diff
@@ -839,6 +839,11 @@ local_traverse(localobject *self, visitproc visit, void *arg)
     return 0;
 }
 
+#define HEAD_LOCK(runtime) \
```
Comment (kumaraditya303) [inline Modules/_threadmodule.c:842]: For 3.12+ we should try to reduce duplication of this macro, will see to it in a separate PR.

**Comment by pablogsal** [comment]
> This race exists in security only branches too, @ambv  how about backporting this to those?

Yeah this should be backported

Excellent job @kumaraditya303 👌

**Review by pablogsal — APPROVED** [review]

File: Modules/_threadmodule.c [hunk 2]
```diff
@@ -849,18 +854,23 @@ local_clear(localobject *self)
...
-            PyObject *v = _PyDict_Pop(tstate->dict, self->key, Py_None);
-            if (v != NULL) {
-                Py_DECREF(v);
-            }
-            else {
-                PyErr_Clear();
+        HEAD_UNLOCK(runtime);
+        while (tstate) {
+            if (tstate->dict) {
+                PyObject *v = _PyDict_Pop(tstate->dict, self->key, Py_None);
```
Comment (pablogsal) [inline Modules/_threadmodule.c:863]: Nit: you can reduce the overall indentation by inverting the check and add a continue but feel free to ignore if you prefer this version

File: Modules/_threadmodule.c [hunk 3]
```diff
@@ -839,6 +839,11 @@ local_traverse(localobject *self, visitproc visit, void *arg)
     return 0;
 }
```
Comment (pablogsal) [inline Modules/_threadmodule.c:840]: Can we add a test to "testcapi"? Even if the tests doesn't answer anything, the sanitizers will detect the problem

File: Modules/_threadmodule.c (hunk 2 above)
Comment (kumaraditya303) [reply Modules/_threadmodule.c:863]: Yeah I went back and forth on this and this looked best to me.

File: Modules/_threadmodule.c (hunk 3 above)
Comment (kumaraditya303) [reply Modules/_threadmodule.c:840]: Added test.

**Comment by kumaraditya303** [comment]
> Excellent job👌

Thanks for the review, I have added a test and verified it, will merge once CI passes.

File: Modules/_threadmodule.c (hunk 1 above)
Comment (kumaraditya303) [reply Modules/_threadmodule.c:842]: https://github.com/python/cpython/pull/100953

**Pull request merged**